File: uis/client/mobile_app/lib/services/remittance_service.dart
import '../models/remittance_transfer.dart';
import '../config/app_config.dart';
import 'api_service.dart';

class RemittanceService {
  final ApiService _apiService;

  RemittanceService(this._apiService);

  static const String _base = AppConfig.remittanceEndpoint;

  Future<List<RemittanceTransfer>> getTransfers({
    int page = 1,
    int limit = 20,
    String? status,
    String? corridor,
  }) async {
    final params = <String, dynamic>{'page': page, 'limit': limit};
    if (status != null) params['status'] = status;
    if (corridor != null) params['corridor'] = corridor;

    final response = await _apiService.get('$_base/v1/transfers', queryParameters: params);
    final data = response.data;
    final rawItems = data is Map ? (data['items'] ?? data['data'] ?? []) : data;
    if (rawItems is! List) return [];
    return rawItems
        .whereType<Map<String, dynamic>>()
        .map(RemittanceTransfer.fromJson)
        .toList();
  }

  Future<RemittanceTransfer> getTransferById(String id) async {
    final response = await _apiService.get('$_base/v1/transfers/$id');
    return RemittanceTransfer.fromJson(Map<String, dynamic>.from(response.data));
  }

  Future<RemittanceTransfer> initiateTransfer({
    required String beneficiaryName,
    required String beneficiaryAccount,
    required String beneficiaryBank,
    required double amount,
    required String senderCurrency,
    required String beneficiaryCurrency,
    required String corridor,
    String? narration,
  }) async {
    final response = await _apiService.post('$_base/v1/transfers', data: {
      'beneficiary_name': beneficiaryName,
      'beneficiary_account': beneficiaryAccount,
      'beneficiary_bank': beneficiaryBank,
      'amount': amount,
      'sender_currency': senderCurrency,
      'beneficiary_currency': beneficiaryCurrency,
      'corridor': corridor,
      if (narration != null) 'narration': narration,
    });
    return RemittanceTransfer.fromJson(Map<String, dynamic>.from(response.data));
  }

  Future<Map<String, dynamic>> getExchangeRate({
    required String fromCurrency,
    required String toCurrency,
    required String corridor,
  }) async {
    final response = await _apiService.get('$_base/v1/exchange-rates', queryParameters: {
      'from': fromCurrency,
      'to': toCurrency,
      'corridor': corridor,
    });
    return Map<String, dynamic>.from(response.data);
  }

  Future<List<Map<String, dynamic>>> getCorridors() async {
    final response = await _apiService.get('$_base/v1/corridors');
    final data = response.data;
    final rawItems = data is Map ? (data['items'] ?? data['data'] ?? []) : data;
    if (rawItems is! List) return [];
    return rawItems.whereType<Map<String, dynamic>>().toList();
  }

  Future<Map<String, dynamic>> getStats({String? from, String? to}) async {
    final params = <String, dynamic>{};
    if (from != null) params['from'] = from;
    if (to != null) params['to'] = to;
    final response = await _apiService.get('$_base/v1/stats', queryParameters: params);
    return Map<String, dynamic>.from(response.data);
  }

  Future<List<Map<String, dynamic>>> getBeneficiaries() async {
    final response = await _apiService.get('$_base/v1/beneficiaries');
    final data = response.data;
    final rawItems = data is Map ? (data['items'] ?? data['data'] ?? []) : data;
    if (rawItems is! List) return [];
    return rawItems.whereType<Map<String, dynamic>>().toList();
  }
}

File: uis/client/web2/src/services/remittance_service.ts
import { apiService } from './api_service';

const BASE = '/remittance/v1';

export interface RemittanceTransfer {
  id: string;
  sender_name?: string;
  beneficiary_name?: string;
  amount: number | string;
  currency?: string;
  corridor?: string;
  status?: string;
  created_at?: string;
  fee?: number | string;
  exchange_rate?: number | string;
  [key: string]: unknown;
}

export interface ExchangeRate {
  from_currency: string;
  to_currency: string;
  rate: number;
  updated_at?: string;
}

export interface RemittanceCorridor {
  id: string;
  name: string;
  from_country: string;
  to_country: string;
  currencies: string[];
  enabled: boolean;
}

export interface RemittanceBeneficiary {
  id: string;
  name: string;
  account_number: string;
  bank_name?: string;
  country?: string;
  currency?: string;
}

export interface RemittanceStats {
  total_transfers: number;
  total_volume: number | string;
  completed: number;
  pending: number;
  failed: number;
  [key: string]: unknown;
}

interface ApiResult<T> {
  success: boolean;
  data: T | null;
  message: string;
}

async function call<T>(fn: () => Promise<{ data: unknown }>): Promise<ApiResult<T>> {
  try {
    const res = await fn();
    const payload = res.data as { data?: T; results?: T } | T;
    const data =
      payload && typeof payload === 'object' && ('data' in payload || 'results' in payload)
        ? ((payload as { data?: T; results?: T }).data ?? (payload as { results?: T }).results ?? (payload as T))
        : (payload as T);
    return { success: true, data, message: 'OK' };
  } catch (err: unknown) {
    const msg = err instanceof Error ? err.message : 'Request failed';
    return { success: false, data: null, message: msg };
  }
}

class RemittanceService {
  getTransfers() {
    return call<RemittanceTransfer[]>(() => apiService.get(`${BASE}/transfers`));
  }

  getTransferById(id: string) {
    return call<RemittanceTransfer>(() => apiService.get(`${BASE}/transfers/${id}`));
  }

  initiateTransfer(data: Record<string, unknown>) {
    return call<RemittanceTransfer>(() => apiService.post(`${BASE}/transfers`, data));
  }

  getExchangeRate(fromCurrency: string, toCurrency: string) {
    return call<ExchangeRate>(() =>
      apiService.get(`${BASE}/exchange-rates`, { from_currency: fromCurrency, to_currency: toCurrency })
    );
  }

  getCorridors() {
    return call<RemittanceCorridor[]>(() => apiService.get(`${BASE}/corridors`));
  }

  getStats() {
    return call<RemittanceStats>(() => apiService.get(`${BASE}/stats`));
  }

  getBeneficiaries() {
    return call<RemittanceBeneficiary[]>(() => apiService.get(`${BASE}/beneficiaries`));
  }
}

export const remittanceService = new RemittanceService();
